runtime/jailer.go

```diff
@@ -0,0 +1,403 @@
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/firecracker-microvm/firecracker-go-sdk"
+
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+)
+
+const (
+	kernelImageFileName   = "kernel-image"
+	jailerHandlerName     = "firecracker-containerd-jail-handler"
+	jailerFifoHandlerName = "firecracker-containerd-jail-fifo-handler"
+	runcConfigPath        = "/etc/containerd/firecracker-runc-config.json"
+	rootfsFolder          = "rootfs"
+
+	// TODO evenetually we can get rid of this when we add usernamespaces to
+	// jailing.
+	jailerUID = 9999
+	jailerGID = 9999
+)
+
+// newJailer is a constructor for the given jailers. By default, the jailer is turned off.
+func newJailer(ctx context.Context, logger *logrus.Entry, ociBundlePath string, config *Config, request *proto.CreateVMRequest) (jailer, error) {
+	if request == nil {
+		return noopJailer{}, nil
+	}
+
+	if request.JailerConfig == nil {
+		return noopJailer{}, nil
+	}
+
+	cfg := runcJailConfig{
+		OCIBundlePath:  ociBundlePath,
+		UID:            jailerUID,
+		GID:            jailerGID,
+		RuncBinaryPath: config.JailerConfig.RuncBinaryPath,
+	}
+
+	return newRuncJailer(logger, cfg), nil
+}
+
+// jailer will allow modification and provide options to the the Firecracker VM
+// to allow for jailing. In addition, this will allow for given files to be exposed
+// to the jailed filesystem.
+type jailer interface {
+	// BuildJailedMachine will modify the firecracker.Config and provide
+	// firecracker.Opt to be passed into firecracker.NewMachine which will allow
+	// for the VM to be jailed.
+	BuildJailedMachine(*Config, *firecracker.Config, string) ([]firecracker.Opt, error)
+	// ExposeDeviceToJail will expose the given device provided by the snapshotter
+	// to the jailed filesystem
+	ExposeDeviceToJail(string) error
+}
+
+// noopJailer is a jailer that returns only successful responses and performs
+// no operations during calls
+type noopJailer struct{}
+
+func (j noopJailer) BuildJailedMachine(cfg *Config, machineConfig *firecracker.Config, vmID string) ([]firecracker.Opt, error) {
+	return []firecracker.Opt{}, nil
+}
+
+func (j noopJailer) ExposeDeviceToJail(path string) error {
+	return nil
+}
+
+type runcJailConfig struct {
+	OCIBundlePath  string
+	UID            uint32
+	GID            uint32
+	RuncBinaryPath string
+}
+
+// runcJailer uses runc to set up a jailed environment for the Firecracker VM.
+type runcJailer struct {
+	logger *logrus.Entry
+
+	Config runcJailConfig
+}
+
+func newRuncJailer(logger *logrus.Entry, config runcJailConfig) *runcJailer {
```

Comment:
From a code structure point of view, I think I'd prefer to see the different jailer implementations split out into their own files, e.g. `noop_jailer.go` and `runc_jailer.go`, with only the common code in `jailer.go`.